Follow symbolic links when recording artifacts

lukpueh opened the pull request:
**Description of the changes being introduced by the pull request**:
This PR adds an optional parameter to the artifact record function to toggle following symbolic links to directories.

Furthermore the PR modifies the calling `runlib` functions, used with `in-toto-run` and `in-toto-record`,  to call with `follow_links=True` .

*UPDATE:*
This change also affects in-toto's final product verification, where, when running inspections, we record artifacts (now following symlinks) in the directory, where the verification is executed.

**Please verify and check that the pull request fulfills the following
requirements**:

- [ ] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [ ] Tests have been added for the bug fix or new feature
- [ ] Docs have been added for the bug fix or new feature

Commit:
Add follow_links parameter to artifact recording

record_artifacts_as_dict uses Python's os.walk to find files
to record. This commit adds an optional boolean parameter handed
through to os.walk that toggles following symlinks (default is
false).

Note that this parameter only concerns directories, linked files
however are always recorded, independently of this parameter.

Commit:
Change in-toto-run/record to follow symbolic

This commit changes runlib's functions used to carry out in-toto
steps, creating link metadata, to follow any symbolic links to
directories when recording artifacts.

Commit:
Add/updates record artifact tests for symlinks

Add more symlink tests, testing the new follow_links parameter
to record_artifacts_as_dict with symlinks to files, dirs and
dead symlinks.

JustinCappos commented:
Are there security implications for this?  Symlinks are very useful in causing a lot of race conditions.

lukpueh commented:
Good question. Do you have anything specific in mind? I am not sure if race conditions are a problem, as we only read from files. The only potential issue that I can think of is getting trapped in an infinite recursion (see [note in `os.walk` docs](https://docs.python.org/3.6/library/os.html#os.walk)).

Btw. I forgot to mention that the change also affects in-toto's final product verification, where, when running inspections, we also record artifacts in the directory, where the verification is executed.

vladimir-v-diaz commented on in_toto/runlib.py:
```diff
@@ -87,7 +87,7 @@ def _apply_exclude_patterns(names, exclude_patterns):
 
 
-def record_artifacts_as_dict(artifacts):
+def record_artifacts_as_dict(artifacts, follow_links=False):
```
Unrelated to this pull request:

> Paths are normalized for matching and storing by left stripping "./"

Is this behavior similar to https://github.com/theupdateframework/tuf/issues/639?  @JustinCappos recommended that we raise an exception if a path includes a leading "/"

lukpueh replied:
No, this is different. It is about normalizing/unifying relative path that might start with or without `./`
```python
>>> os.path.normpath("./relative/to/cwd") == "relative/to/cwd"
True
```
IIRC, your case is about normalizing/unifying absolute and relative paths, when the absolute root is also the relative root of a path, right?

vladimir-v-diaz commented on in_toto/runlib.py:
```diff
@@ -130,6 +130,11 @@ def record_artifacts_as_dict(artifacts):
 
+    follow_links: (optional)
+            Follow symbolic links if the linked dir exists (default is False).
+            NOTE: This parameter toggles following linked directories only,
```
> Follow symbolic links if the **linked** dir exists (default is False).

I recommend you use "symlink" when referring to symbolic links, otherwise it clashes with in-toto's use of "link cmd," "link files," "Link Object," etc.

vladimir-v-diaz replied:
Would it help to rename `follow_links` to something like `follow_symlink_dir`?  It might clear up the possible confusion with the default/different behavior with symlinked files.

vladimir-v-diaz replied:
You should also consider adding a note about an infinite recursion possibility if a symlink points to a parent directory of itself.

lukpueh replied:
All good points, I will change as you suggest

vladimir-v-diaz commented on in_toto/runlib.py:
```diff
@@ -398,7 +403,7 @@ def in_toto_run(name, material_list, product_list, link_cmd_args,
     log.info("Recording materials '{}'...".format(", ".join(material_list)))
-  materials_dict = record_artifacts_as_dict(material_list)
+  materials_dict = record_artifacts_as_dict(material_list, follow_links=True)
 
```
So `follow_links` is always set to true for someone using/calling `in_toto_run`?  The docstring for `in_toto_run` should probably mention this default behavior.

lukpueh replied:
Will do

vladimir-v-diaz reviewed (commented):
Here is my review of `runlib.py.`

vladimir-v-diaz commented on tests/test_runlib.py:
```diff
@@ -144,7 +142,7 @@ def test_bad_base_path_setting(self):
       in_toto.settings.ARTIFACT_BASE_PATH = setting
       with self.assertRaises(in_toto.exceptions.SettingsError):
-        record_artifacts_as_dict(["."])
+        record_artifacts_as_dict(["."] )
```
Why was a space added?

vladimir-v-diaz commented on tests/test_runlib.py:
```diff
@@ -181,8 +179,98 @@ def test_record_dot_check_files_hash_dict_schema(self):
+
+
+  def test_record_follow_symlinked_directories(self):
+    """Record files in symlinked dirs if follow_links is True. """
```
I don't recall whether the docstring for `follow_links` explains it, but which file path is recorded when a symlinked directory is visited?  The original path, or the resolved one?

vladimir-v-diaz reviewed (commented):
Here is my review of `test_runlib.py.`

Commit:
Add note about symlinks note to run/record cli

Per default symlinks are followed when recording in-toto metadata
using in-toto-run or in-toto-record. This commit adds a note to
the cli tools help messages for materials and products arguments
and adapts docstrings.

Commit:
Remove bogus whitespace in function call

Commit:
Rename follow_links to follow_symlink_dirs

As suggested by @vlad-v-diaz the new name clarifies that the
option only toggles following symlinked dirs when recording
artifacts. Symlinked files are always recorded, independently of
this option.

Commit:
Update docstring for follow_symlink_dirs

Mention that the original link path is recorded, not the resolved
one and point out the infinite recursion hazard.

lukpueh commented:
Thanks for the review @vladimir-v-diaz. Merging...
